### PR TITLE
add dateonly and timeonly support

### PR DESCRIPTION
**Pull request by SimonCropp**

### CODE REVIEW

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]:
```suggestion
            DateOnly _ => value, // no boxing
```
(to prevent boxing)

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:35]:
```suggestion
            TimeOnly _ => value, // no boxing
```
(to prevent boxing)

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:38]:
```suggestion
            _ => ThrowValueConversionError(value)
```

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:38]:
```suggestion
            _ => ThrowValueConversionError(value)
```

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:44]:
```suggestion
            TimeOnly d => d.ToString("HH:mm:ss.FFFFFFF", DateTimeFormatInfo.InvariantInfo),
```
So that fractional parts of a second are not serialized if they are zero.

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:51]:
```suggestion
            if (TimeOnly.TryParseExact(stringValue, "HH:mm:ss.FFFFFFF", DateTimeFormatInfo.InvariantInfo, DateTimeStyles.None, out var date))
```

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:56]:
```suggestion
            throw new FormatException($"Could not parse time. Expected HH:mm:ss.FFFFFFF. Value: {stringValue}");
```

**Review by Shane32 — CHANGES_REQUESTED** [review]

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (Shane32) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:21]:
```suggestion
                "[ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) standard. Format is `HH:mm:ss.FFFFFFF`.";
```
I missed this one. For consistency let’s change it too. Otherwise looks good.

**Review by Shane32 — CHANGES_REQUESTED** [review]

**Comment by SimonCropp** [comment]
seems the LGTM step doesnt support net6?

**Comment by Shane32** [comment]
> seems the LGTM step doesnt support net6?

No

**Review by Shane32 — APPROVED** [review]

**Comment by Shane32** [comment]
Note that we can’t issue a new release until @joemcbride updates the nuget tokens

File: src/GraphQL/GraphQL.csproj
Comment (sungam3r) [inline src/GraphQL/GraphQL.csproj:4]:
```suggestion
    <TargetFrameworks>netstandard2.1;netstandard2.0;net6</TargetFrameworks>
```

to align with other usages

File: src/GraphQL/Conversion/ValueConverter.cs
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs:41]: No DateTimeStyles ?

Comment (SimonCropp) [reply src/GraphQL/Conversion/ValueConverter.cs:41]: dateonly has no timezone, hence no universal time, so `DateTimeStyles.AdjustToUniversal | DateTimeStyles.AssumeUniversal` doesnt make sense. should i have included other DateTimeStyles?

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (sungam3r) [inline src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: What about `DateTime` support here? Also what about supporting `DateOnly` for `DateGraphType`? I think that `DateGraphType` and `DateOnlyGraphType` now are some sort of twins.

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (sungam3r) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:11]:
```suggestion
    /// Format is `HH:mm:ss.FFFFFFF`.
```

File: src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs
Comment (sungam3r) [inline src/GraphQL/Types/Scalars/TimeOnlyGraphType.cs:53]:
```suggestion
            if (TimeOnly.TryParseExact(stringValue, "HH:mm:ss.FFFFFFF", DateTimeFormatInfo.InvariantInfo, DateTimeStyles.None, out var time))
            {
                return time;
```

File: src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs
Comment (sungam3r) [inline src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:46]: Please add more tests with "01:02:03.00400009" and "01:02:03.0040009" values.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (sungam3r) [inline src/GraphQL.ApiTests/GraphQL.approved.txt:578]: .NET 6 TFM side effect ? I think that in case of ifdefs there should be multiple approval files - for each TFM.

File: src/GraphQL/Conversion/ValueConverter.cs
Comment (sungam3r) [reply src/GraphQL/Conversion/ValueConverter.cs:41]: I don't know. I just see that `Parse` method accepts `DateTimeStyles` argument.

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (Shane32) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: I would not parse DateTime here. If someone receives DateTime that would be a coding error. We don’t want to allow people to write errors in new code.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I wouldn’t. It will just make it more painful to update the approval files. And then whenever there is an api change, there will be three approval files to review. The approvals are supposed to help us, not annoy us.

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (sungam3r) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: Why coding error?

Comment (sungam3r) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: This is only a matter of ease of use. It seems to me quite natural that this scalar type will support several .NET types. Semantics is not broken.

**Comment by joemcbride** [comment]
> Note that we can’t issue a new release until @joemcbride updates the nuget tokens

I thought I had done this ... will double check.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: First of all, it should reflect reality. Now it shows public API for NET6 only. As soon as we have different APIs for different TFM, we actually **have to** have different approval files. 

> The approvals are supposed to help us, not annoy us.

Agree. That is why we need to make different approval files in order not to get confused in subsequent independent changes for netstandard/NET TFMs.

**Comment by joemcbride** [comment]
> I thought I had done this ... will double check.

Yes I updated the token in May.  I went ahead and regenerated it again and updated the Org secret.  `NUGET_AUTH_TOKEN`

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (Shane32) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: No, the scalar type should match the variable type.  There is absolutely no reason for the DateOnly scalar type to support any other .NET type as a variable.  If someone can think of a legitimate reason, we can change it upon request.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: No, we don't.  The existing approval files are sufficient.  There is no confusion.  Nobody looks at the approval files to know what APIs are supported.  It's just a way for us to review changes to the public API base to be sure we don't issue a breaking change.  It's not a published document that shows that for such-and-such a TFM there are specific public members.

**Comment by Shane32** [comment]
> > I thought I had done this ... will double check.
> 
> Yes I updated the token in May. I went ahead and regenerated it again and updated the Org secret. `NUGET_AUTH_TOKEN`

Thank you!!

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I did not say that this is a public document.

>  Nobody looks at the approval files to know what APIs are supported for a specific TFM.

**We** look to know about breaking changes. It's for us of course. Imagine one adds `TimeOnlyGraphType` in netstandard2.0 TFM in v4.8. Then for some reason this type will be removed in v4.9. It's a breaking change obviously but we will not even notice this change, it will be hidden from our eyes because now we see only public API changes for NET6.

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (sungam3r) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: > No, the scalar type should match the variable type. 

What means _match_ ?
изображение

Comment (sungam3r) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: `IntGraphType` accepts `ushort` variable thought the is "dedicated" `UShortGraphType`. The same is true for a plethora of other types.

Comment (Shane32) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: Right, because it is not known what System.Text.Json will deserialize a number into.  In this case the value stored as json is a string.  It is not going to be DateTime, DateOnly, or any other .NET type.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I do not think this "issue" is worth the considerable trouble to implement.

Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I will try.

Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: #2730

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (Shane32) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: Regardless, does not mean that DateTime should be coerced to DateOnly.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (Shane32) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: Perhaps you misunderstood me.  I was not talking about the effort to implement multiple approval files.  I was talking about the effort keep them up to date and review them on an ongoing basis in the future.

File: src/GraphQL/Conversion/ValueConverter.cs
Comment (SimonCropp) [reply src/GraphQL/Conversion/ValueConverter.cs:41]: agreed. also i cant find any docs about how DateTimeStyles works in the context of DateOnly

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (SimonCropp) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: i would prefer that DateGraphType only supported dateonly and DateOnlyGraphType would not need to exist. but i could not work out how to do that gracefully. if it was my project i would drop support for older fwks and only support net6

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (SimonCropp) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: if we were using Verify i would use a  `UniqueFor`  to produce one approved file per 
 target fwk https://github.com/VerifyTests/Verify/blob/main/docs/naming.md#uniquefor 

> Perhaps you misunderstood me. I was not talking about the effort to implement multiple approval files. I was talking about the effort keep them up to date and review them on an ongoing basis in the future.

verify supports bulk approval (through a system wide kb shortcut or a UI)

Comment (SimonCropp) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: but IMO this topic should not block this PR. i can be handled in a diff issue/pr

File: src/GraphQL/Types/Scalars/DateOnlyGraphType.cs
Comment (sungam3r) [reply src/GraphQL/Types/Scalars/DateOnlyGraphType.cs:35]: Well, it will be completely radically 😄 . So... I give up, let it remain as it is.

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I fixed approval files bloat and `AllowNull` attributes side effect as well. See #2730 again, please.

Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: I suggest merge this PR after #2730

File: src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs
Comment (SimonCropp) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:46]: good catch. only 3 points of milliseconds are supported. i will fix all the FFFFFs https://docs.microsoft.com/en-us/dotnet/api/system.timeonly.-ctor?view=net-6.0#System_TimeOnly__ctor_System_Int32_System_Int32_System_Int32_System_Int32_

File: src/GraphQL/Conversion/ValueConverter.cs
Comment (sungam3r) [reply src/GraphQL/Conversion/ValueConverter.cs:41]: So maybe use `DateTimeStyles.None` explicitly to highlight this decision and go on?

File: src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs
Comment (Shane32) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:46]: What do you mean?  This prints 00:00:00.0000001

```cs
Console.WriteLine(new TimeOnly(1).ToString("HH:mm:ss.FFFFFFF"));
```

TimeOnly stores times in ticks: "A time of day expressed in the number of 100-nanosecond units since 00:00:00.0000000."

Comment (sungam3r) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:46]: WOW!

Comment (SimonCropp) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:46]: sorry. i crossed the streams. ignore my previous comment

File: src/GraphQL.ApiTests/GraphQL.approved.txt
Comment (sungam3r) [reply src/GraphQL.ApiTests/GraphQL.approved.txt:578]: OK, this goes first :)

**Comment by sungam3r** [comment]
@SimonCropp Waiting for FFFFFFF tests. Otherwize looks good.

File: src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs
Comment (SimonCropp) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:50]: so the difficulty is the readable timeonly constructors do not support anything smaller than milliseconds. u need to use tick overload, resulting in 

```
_type.ParseValue("01:02:03.0040009").ShouldBe(new TimeOnly(37230040009));
```

ugly but it works

**Review by sungam3r — APPROVED** [review]

**Comment by SimonCropp** [comment]
@sungam3r 

> Waiting for FFFFFFF tests. Otherwize looks good.

pushed. let me know if it looks good

File: src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs
Comment (sungam3r) [reply src/GraphQL.Tests/Types/TimeOnlyGraphTypeTests.cs:50]: ok

**Comment by SimonCropp** [comment]
awesome. any chance of getting a beta nuget pushed so i can smoke test?

**Comment by sungam3r** [comment]
We use GitHub package registry for preview packages - https://github.com/orgs/graphql-dotnet/packages?repo_name=graphql-dotnet

**Comment by SimonCropp** [comment]
@sungam3r so there should be a 4.7 beta? 
image

**Comment by sungam3r** [comment]
These versions are from `Directory.Build.props`. We had no `4.7.-preview` version here (but we SHOULD, we just forgot to bump version). Only `4.6.0-preview` for `master` and `5.0.0-preview` for `develop` now. Just pull the latest `4.6.0-preview` with your changes.

**Pull request merged**
